### PR TITLE
Added quiet mode for CLI

### LINKED ISSUE #64: Quiet mode for slang-cli 'run' command
'run' currently prints out volumes of extra information that would be very awkward to use as part of a script, or anything else that parses results, eg:

```text
$ ../score-lang-cli/slang/bin/slang run division.sl --i input1=10,input2=-1
Database name is: H2 databaseMajorVersion is: 1
- divider
- printer
10/-1 = -10.0
Flow : division finished with result : SUCCESS
Jan 28, 2015 6:00:22 PM org.springframework.shell.core.AbstractShell handleExecutionResult
INFO: Flow execution time took  0:00:03.158 , with execution id : 101600001
```
A quiet mode, that only returns what is specified as output by the flow would be very useful.  In this case, that would be:
```
10/-1 = -10.0
```
Additionally (and I'll file a separate issue on this), it would be a major plus if the exit code reflected the success or failure of the flow.

### PR DESCRIPTION
**Pull request by tudorlesan**
Signed-off-by: Lesan Tudor <tudor-andrei.lesan@hp.com>
Fixes #64

### CODE REVIEW

File: score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java [hunk 1]
```diff
@@ -64,7 +64,8 @@ public String run(
             @CliOption(key = {"", "f", "file"}, mandatory = true, help = "Path to filename. e.g. slang run --f C:\\Slang\\flow.yaml") final File file,
             @CliOption(key = {"cp", "classpath"}, mandatory = false, help = "Classpath , a directory comma separated list to flow dependencies, by default it will take flow file dir") final List<String> classPath,
             @CliOption(key = {"i", "inputs"}, mandatory = false, help = "inputs in a key=value comma separated list") final Map<String, Serializable> inputs,
-            @CliOption(key = {"spf", "system-property-file"}, mandatory = false, help = "comma separated list of system property file locations") final List<String> systemPropertyFiles) throws IOException {
+            @CliOption(key = {"spf", "system-property-file"}, mandatory = false, help = "comma separated list of system property file locations") final List<String> systemPropertyFiles,
```
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java]: what is this related too?

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServices.java [hunk 2]
```diff
@@ -24,4 +24,5 @@
     void subscribe(ScoreEventListener eventHandler, Set<String> eventTypes);
     Long trigger(CompilationArtifact compilationArtifact, Map<String, ? extends Serializable> inputs, Map<String, ? extends Serializable> systemProperties);
     Long triggerSync(CompilationArtifact compilationArtifact, Map<String, ? extends Serializable> inputs, Map<String, ? extends Serializable> systemProperties);
+    Long triggerSync(CompilationArtifact compilationArtifact, Map<String, ? extends Serializable> inputs, Map<String, ? extends Serializable> systemProperties, boolean quiet);
```
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServices.java]: why new method?

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java [hunk 3]
```diff
@@ -95,11 +96,47 @@ public Long triggerSync(CompilationArtifact compilationArtifact, Map<String, ? e
         slang.unSubscribeOnEvents(scoreEventListener);
         return executionId;
     }
+    @Override
+    public Long triggerSync(CompilationArtifact compilationArtifact, Map<String, ? extends Serializable> inputs, Map<String, ? extends Serializable> systemProperties, boolean isQuiet){
```
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: this is code duplication.. why not add the param quiet to existing method?

File: score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java (hunk 1 above)
Comment (TudorLesanHP) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java]: don't know why this appeared. Didn't intend to.

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServices.java (hunk 2 above)
Comment (TudorLesanHP) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServices.java]: I meant to delete the old one. I didn't have time to check if everything still works fine so I overloaded the method.

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java (hunk 3 above)
Comment (TudorLesanHP) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: again, I overloaded the method to make sure everything still works before checking it. I will make sure there only one method

File: score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java (hunk 1 above)
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java]: again.. update from master...

**Comment by meirwah** [comment]
@TudorLesanHP please fix issues and add test before pull request..

**Comment by meirwah** [comment]
@orius123 or @avimora  can you review this too?

**Comment by TudorLesanHP** [comment]
fixes #64

File: score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java [hunk 4]
```diff
@@ -72,9 +73,15 @@ public String run(
...
+            if(logValue == null || logValue.equals("DEBUG")) {
+                id = scoreServices.triggerSync(compilationArtifact, inputs, systemProperties, false);
+            }else if(logValue.equals("INFO")){
+                id = scoreServices.triggerSync(compilationArtifact, inputs, systemProperties, true);
+            }else{
+                return "Wrong value for --log. Valid: INFO, DEBUG.";
+            }
             stopWatch.stop();
-            return triggerSyncMsg(id, stopWatch.toString());
+            return (logValue == null || logValue.equals("DEBUG")) ? triggerAsyncMsg(id, compilationArtifact.getExecutionPlan().getName()) : "";
```
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java]: i do not like the complexity here.. there should be only -Q param in my opinion
it is either quiet or not.. let's not make things complex for no reason

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java [hunk 5]
```diff
@@ -131,12 +143,43 @@ public synchronized void onEvent(ScoreEvent scoreEvent) throws InterruptedExcept
...
                 case EVENT_EXECUTION_FINISHED :
+                   if (!flowFinished.get()){
+                        flowFinished.set(true);
+                    }
                     printFinishEvent(data);
                     break;
             }
         }
 
+        private void printFlowOutputs(Map<String, Serializable> data) {
```
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: why is this so complex???

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java [hunk 6]
```diff
@@ -131,12 +143,43 @@ public synchronized void onEvent(ScoreEvent scoreEvent) throws InterruptedExcept
                         printWithColor(Ansi.Color.YELLOW, prefix + taskName);
                     }
                     break;
+                case ScoreLangConstants.EVENT_OUTPUT_END:
+                    if(quietListener){
```
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: isn't the the way around?

File: score-lang-cli/src/test/java/org/openscore/lang/cli/SlangCLITest.java [hunk 7]
```diff
@@ -45,9 +45,9 @@
 public class SlangCLITest {
 
     private final static String[] CONTEXT_PATH = { "classpath*:/META-INF/spring/test-spring-shell-plugin.xml" };
-    private final static String FLOW_PATH_SLAH = "C:\\\\flow.yaml";
+    private final static String FLOW_PATH_BACKSLASH_INPUT = "C:\\\\flow.yaml";
```
Comment (meirwah) [inline score-lang-cli/src/test/java/org/openscore/lang/cli/SlangCLITest.java]: why was this changed?

Comment (TudorLesanHP) [inline score-lang-cli/src/test/java/org/openscore/lang/cli/SlangCLITest.java]: the variable name was not consistent since the value contained backslash not forwardslash

Comment (meirwah) [inline score-lang-cli/src/test/java/org/openscore/lang/cli/SlangCLITest.java]: ok

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java (hunk 6 above)
Comment (TudorLesanHP) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: the quiet mode refers to not showing tasks that are executed. The flow outputs are printed even if quietListener is true. atm the implementation is: normal mode - print tasks (no outputs), quiet mode - print flow outputs (no tasks)
Once we merge the flow_outputs_to_cli branch the normal mode will print flow outputs by default and the condition won't be necessary.

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java (hunk 5 above)
Comment (TudorLesanHP) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: this gets the outputs of the flow, truncates any value to length 30 , puts "(empty)" in empty values and replaces newlines with space. it isn't so complex, what do you think I could simplify/remove?

File: score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java (hunk 4 above)
Comment (TudorLesanHP) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java]: I agree that the -Q option would be simpler, but as far as I know, SpringShell doesn't offer us this posibility

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java (hunk 5 above)
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: don't hard code numbrers like 30, use StringUtils to simplify ssome stuff here..

File: score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java (hunk 4 above)
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java]: @CliOption(key = {"", "q", "quiet"}, mandatory = false, help = "quiet", specifiedDefaultValue = "true",unspecifiedDefaultValue = "false") final Boolean quiet

**Comment by TudorLesanHP** [comment]
@meirwah fixed

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java [hunk 8]
```diff
@@ -106,37 +117,70 @@ public boolean isFlowFinished() {
...
+                        String prefix = getPrefix(data);
                         printWithColor(Ansi.Color.YELLOW, prefix + taskName);
                     }
                     break;
-                case EVENT_EXECUTION_FINISHED :
+                case ScoreLangConstants.EVENT_OUTPUT_END:
+                        printOutputs(data);
+                    break;
+                case EVENT_EXECUTION_FINISHED:
+                    if (!flowFinished.get()) {
```
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: why was this added?
how does it relate to the issue?

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java [hunk 9]
```diff
@@ -106,37 +117,70 @@ public boolean isFlowFinished() {
...
+                        flowFinished.set(true);
+                    }
                     printFinishEvent(data);
                     break;
             }
         }
 
+        private void printOutputs(Map<String, Serializable> data) {
```
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: why on a fix for quiet mode, you have changes to the outputs print method?

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java (hunk 8 above)
Comment (TudorLesanHP) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: "A quiet mode, that only returns what is specified as output by the flow would be very useful". But when I first wrote it, the printOutputs method for the quiet_mode_for_cli branch was different than the printOutputs method for the outputs_in_cli branch since one printed just flow outputs and the other also printed step outputs. Now they are the same, so the code for the method is on both branches.

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java (hunk 9 above)
Comment (TudorLesanHP) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: both #5 and #64 required outputs to be printed on the console. I first wrote the printOutputs method for #5 which printed all outputs including step outputs. Then I wrote the printOutputs for #64 which only printed flow outputs. Now since I changed the one on #5 both methods are the same on both branches.

File: score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java (hunk 1 above)
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java]: it is still here, @TudorLesanHP  are you going to update from master?

Comment (TudorLesanHP) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java]: I already did. No conflict appeared. master doesn't even appear any more on the "Merge branches"

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java (hunk 8 above)
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: @TudorLesanHP that is a bad apporach, each branch should hold ONLY changes relevent to what the bug will fix, after you merge one of them, you will go to the other branch, and fix accordingly, what you are doing is not by the Git bast practices...

File: score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java [hunk 10]
```diff
@@ -72,9 +73,9 @@ public String run(
         if (!triggerAsync) {
             StopWatch stopWatch = new StopWatch();
             stopWatch.start();
-            id = scoreServices.triggerSync(compilationArtifact, inputs, systemProperties);
+            id = scoreServices.triggerSync(compilationArtifact, inputs, systemProperties, quiet);
             stopWatch.stop();
-            return triggerSyncMsg(id, stopWatch.toString());
+            return (!quiet) ? triggerAsyncMsg(id, compilationArtifact.getExecutionPlan().getName()) : "";
```
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java]: this looks better (less verbose):
quiet ? StringUtils.EMPTY : triggerAsyncMsg(id, compilationArtifact.getExecutionPlan().getName())

Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java]: you changed the method name from triggerSyncMsg to triggerAsyncMsg ? why? it's to diff methods?

File: score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java [hunk 11]
```diff
@@ -72,9 +73,9 @@ public String run(
...
         }
         id = scoreServices.trigger(compilationArtifact, inputs, systemProperties);
         return triggerAsyncMsg(id, compilationArtifact.getExecutionPlan().getName());
```
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java]: so the Async method does not change?? why?

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java (hunk 8 above)
Comment (TudorLesanHP) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: I know. The changes were unique for each branch. I first did #5, then #64. I understood that each issue should have different outputs being printed. But now they contain the same logic after the changes to #5. After you merge one of the branches. I will go to the other one and fix any conflicts.

File: score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java (hunk 10 above)
Comment (TudorLesanHP) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java]: my mistake. Sorry about that

File: score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java (hunk 11 above)
Comment (TudorLesanHP) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java]: I only added the option for the synchronous execution. You said in #64 that the asynchronous execution doesn't print tasknames, so quiet mode wouldn't affect it, would it?

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java [hunk 12]
```diff
@@ -69,18 +71,27 @@ public Long trigger(CompilationArtifact compilationArtifact, Map<String, ? exten
...
         //add start event
         Set<String> handlerTypes = new HashSet<>();
-        handlerTypes.add(EventConstants.SCORE_FINISHED_EVENT);
-        handlerTypes.add(EventConstants.SCORE_ERROR_EVENT);
-        handlerTypes.add(EventConstants.SCORE_FAILURE_EVENT);
-        handlerTypes.add(SLANG_EXECUTION_EXCEPTION);
-        handlerTypes.add(EVENT_EXECUTION_FINISHED);
-        handlerTypes.add(EVENT_INPUT_END);
+        if(isQuiet){
+            handlerTypes.add(EVENT_OUTPUT_END);
```
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: why was this added? does it mean that step outputs are always printed?

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java (hunk 9 above)
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: please remove the ouputs changes that are unrelated..
this nmakes it impossible to track the logic of the change..

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java (hunk 12 above)
Comment (TudorLesanHP) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: This just specifies what events are tracked in both modes. Quiet mode only tracks outputs and the finish event while normal mode also tracks all others. The printOutputs method checks so that only the flow outputs get printed. So step outputs are never printed. Should I also add the exception/error events to be tracked for quiet mode? ATM no errors are printed to CLI in quiet mode.

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java (hunk 9 above)
Comment (TudorLesanHP) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: #64 requires flow outputs to be printed (#5). Maybe first merge #5 to master then I can remove the output changes on #64 then I can get them from the master?

**Comment by meirwah** [comment]
@orius123 please review this too...

**Comment by meirwah** [comment]
more tests are needed here.. 
where are tasts for the async?

**Comment by meirwah** [comment]
@tudorlesan please update this branch...
@orius123 / @avimora  can you review this too?

**Comment by meirwah** [comment]
@avimora can you review this too?

**Comment by meirwah** [comment]
@tudorlesan please update branch...

File: score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java [hunk 13]
```diff
@@ -85,12 +86,12 @@ public String run(
         if (!triggerAsync) {
             StopWatch stopWatch = new StopWatch();
             stopWatch.start();
-            id = scoreServices.triggerSync(compilationArtifact, mergedInputs, systemProperties);
+            id = scoreServices.triggerSync(compilationArtifact, mergedInputs, systemProperties, quiet);
             stopWatch.stop();
-            return triggerSyncMsg(id, stopWatch.toString());
+            return quiet ? StringUtils.EMPTY : triggerSyncMsg(id, compilationArtifact.getExecutionPlan().getName());
         }
```
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java]: compilationArtifact.getExecutionPlan().getName() for the duration param????
that does not make any sense....

Comment (tudorlesan) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java]: sorry, do not know how that got in there. will fix it asap

Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java]: Please add test to verify it works...

Comment (tudorlesan) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java]: Fixed this. I already added tests for quiet mode. This just printed the flow name instead of the execution duration, but I can't really test this since the duration could be different for different systems.

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/SyncTriggerEventListener.java [hunk 14]
```diff
@@ -93,6 +93,7 @@ public synchronized void onEvent(ScoreEvent scoreEvent) throws InterruptedExcept
                 break;
 
             case EVENT_EXECUTION_FINISHED :
+                flowFinished.set(true);
```
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/SyncTriggerEventListener.java]: still not sure it is really needed... @orius123  WDYT?

Comment (orius123) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/SyncTriggerEventListener.java]: On the contrary, I think we should stop listening to score finished event. And only to slang event.

**Comment by meirwah** [comment]
so :+1: to merge this...

**Comment by meirwah** [comment]
fixes #64

File: score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java [hunk 15]
```diff
@@ -85,12 +86,12 @@ public String run(
...
             stopWatch.start();
-            id = scoreServices.triggerSync(compilationArtifact, mergedInputs, systemProperties);
+            id = scoreServices.triggerSync(compilationArtifact, mergedInputs, systemProperties, quiet);
             stopWatch.stop();
-            return triggerSyncMsg(id, stopWatch.toString());
+            return quiet ? StringUtils.EMPTY : triggerSyncMsg(id, stopWatch.toString());
         }
-        id = scoreServices.trigger(compilationArtifact, mergedInputs, systemProperties);
-        return triggerAsyncMsg(id, compilationArtifact.getExecutionPlan().getName());
+        id = scoreServices.trigger(compilationArtifact, inputs, systemProperties);
```
Comment (tudorlesan) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/SlangCLI.java]: I think here it was mergedInputs instead of inputs. Fixed this in cli_extentions_support
